updater/src/updater.rs

```diff
@@ -255,47 +380,151 @@ impl Updater {
-					fs::copy(&b, &dest).map_err(|e| (format!("Unable to copy update: {:?}", e), true))?;
-					restrict_permissions_owner(&dest, false, true).map_err(|e| (format!("Unable to update permissions: {}", e), true))?;
-					info!(target: "updater", "Installed updated binary to {}", dest.display());
-				}
-				let auto = match self.update_policy.filter {
-					UpdateFilter::All => true,
-					UpdateFilter::Critical if fetched.is_critical /* TODO: or is on a bad fork */ => true,
-					_ => false,
+	fn updater_step(&self) {
+		let current_block_number = self.client.upgrade().map_or(0, |c| c.block_number(BlockId::Latest).unwrap_or(0));
+
+		let mut state = self.state.lock();
+
+		if let Some(latest) = state.latest.clone() {
+			let fetch = |binary| {
+				info!(target: "updater", "Attempting to get parity binary {}", binary);
+				let weak_self = self.weak_self.lock().clone();
+				let latest = latest.clone();
+				let on_fetch = move |res: Result<PathBuf, fetch::Error>| {
+					if let Some(this) = weak_self.upgrade() {
+						let mut state = this.state.lock();
+
+						// Check if the latest release and updater status hasn't changed
+						if state.latest.as_ref() == Some(&latest) {
+							if let UpdaterStatus::Fetching { ref release, binary, retries } = state.status.clone() {
+								match res {
+									// We've successfully fetched the binary
+									Ok(path) => {
+										let setup = |path: &Path| -> Result<(), String> {
+											let dest = this.updates_path(&Self::update_file_name(&release.version));
+											if !dest.exists() {
+												info!(target: "updater", "Fetched latest version ({}) OK to {}", release.version, path.display());
+												fs::create_dir_all(dest.parent().expect("at least one thing pushed; qed")).map_err(|e| format!("Unable to create updates path: {:?}", e))?;
+												fs::copy(path, &dest).map_err(|e| format!("Unable to copy update: {:?}", e))?;
+												restrict_permissions_owner(&dest, false, true).map_err(|e| format!("Unable to update permissions: {}", e))?;
+												info!(target: "updater", "Copied updated binary to {}", dest.display());
+											}
+
+											Ok(())
+										};
+
+										// There was a fatal error setting up the update, disable the updater
+										if let Err(err) = setup(&path) {
+											state.status = UpdaterStatus::Disabled;
+											warn!("{}", err);
+										} else {
+											state.status = UpdaterStatus::Ready { release: release.clone() };
+											// will lock self.state
+											drop(state);
+											this.updater_step();
+										}
+									},
+									// There was an error fetching the update, apply a backoff delay before retrying
+									Err(err) => {
+										let delay = 2usize.pow(retries) as u64;
+										let backoff = (retries, Instant::now() + Duration::from_secs(delay));
```

Comment:
I would suggest putting a maximum here, otherwise it goes from 6 days to 12 days to a month etc around 20 retries, meaning we have like ~2 weeks to fix any issue before it takes a month before any node updates. Having a maximum wait of a week seems pretty reasonable?

Reply:
It's also probably a cause of #8003 crash

Reply:
The current updater logic will drop the backoff as soon as we a new release is pushed, so we won't get stuck in any wait case. I will add a test for that 😉.